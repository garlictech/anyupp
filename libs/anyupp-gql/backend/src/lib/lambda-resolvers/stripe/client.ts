import { Stripe } from 'stripe';

export const createStripeClient = (secret: string) => {
  return new Stripe(secret, {
    apiVersion: '2020-08-27',
  });
};
