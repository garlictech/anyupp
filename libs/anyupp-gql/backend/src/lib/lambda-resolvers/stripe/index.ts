export { createStripeResolvers } from './stripe-resolvers';
export { stripeRequestHandler } from './stripe-request-handler';
export { createStripeWebhookExpressApp } from './stripe-webhook-handler';
export * from './client';
