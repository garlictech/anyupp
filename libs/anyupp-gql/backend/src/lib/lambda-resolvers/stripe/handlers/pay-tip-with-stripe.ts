import * as CrudApi from '@bgap/crud-gql/api';
import Stripe from 'stripe';
import { createTransaction, loadOrder, loadUnit } from '../stripe-graphql-crud';
import { StripeResolverDeps } from '../stripe.utils';
import { loadAndConnectUserForStripe } from './common-stripe';
import { toFixed0Number } from '@bgap/shared/utils';

export const payTipWithStripe =
  (input: CrudApi.PayTipWithStripeInput) =>
  async (
    deps: StripeResolverDeps,
  ): Promise<CrudApi.StartStripePaymentOutput> => {
    const userId = deps.userId;
    console.debug('startStripePayment().start()');

    // 1. Get parameters, orderId and payment method
    const { orderId } = input;
    console.debug('startStripePayment().orderId=' + orderId);

    const order = await loadOrder(orderId)(deps);

    if (!order) {
      throw Error('Order not found with id=' + orderId);
    }

    const paymentMethod = order?.paymentMode.method;
    const paymentMethodId = order?.transaction?.paymentMethodId;

    if (paymentMethod == CrudApi.PaymentMethod.inapp && !paymentMethodId) {
      throw Error(
        'Payment method is missing from request when payment mode is INAPP!',
      );
    }

    if (userId !== order.userId) {
      throw Error(
        "The Order must belongs to the user, the Order's userId field is not match with the logined User's id! Details: order.userId=" +
          order.userId +
          ' vs userId=' +
          userId,
      );
    }

    // 3. Load User
    const createStripeCustomer = paymentMethod == CrudApi.PaymentMethod.inapp;
    const user = await loadAndConnectUserForStripe(
      userId,
      createStripeCustomer,
    )(deps);
    console.debug('startStripePayment().user.loaded=' + user?.id);

    if (!user) {
      throw Error(
        'User initialization failed. User not loaded (and maybe not created)!',
      );
    }

    if (
      paymentMethod == CrudApi.PaymentMethod.inapp &&
      !user.stripeCustomerId
    ) {
      throw Error(
        'User initialization failed. User must have a stripeCustomerId property!',
      );
    }

    // 4. Load unit
    console.debug('startStripePayment().loading unit()=' + order.unitId);
    const unit = await loadUnit(order.unitId)(deps);
    console.debug('startStripePayment().unit().loaded=' + unit?.id);
    if (!unit) {
      throw Error('Unit not found with id=' + order.unitId);
    }

    // 5. Handle INAPP payment
    if (paymentMethod == CrudApi.PaymentMethod.inapp) {
      if (!paymentMethodId) {
        throw Error(
          'Payment method is missing from request when payment mode is INAPP!',
        );
      }

      if (!user.stripeCustomerId) {
        throw Error(
          "User don't have a stripeCustomerId property, but it is mandatory for INAPP mayment mode!",
        );
      }

      const amount = input.tip.value;
      const currency = order.sumPriceShown.currency;

      const stripeAmount: number =
        currency === 'HUF' ? toFixed0Number(amount * 100) : amount;

      // 6a. Create payment intent data
      const paymentIntentData: Stripe.PaymentIntentCreateParams = {
        amount: stripeAmount,
        currency,
        payment_method: paymentMethodId,
        payment_method_types: ['card'],
        customer: user.stripeCustomerId,
        statement_descriptor_suffix: 'tip',
      };

      // 6b. Add optional merchantId to the payment
      if (unit?.merchantId) {
        console.debug('startStripePayment().set merchantId=' + unit.merchantId);
        paymentIntentData.application_fee_amount = 0;
        paymentIntentData.transfer_data = {
          destination: unit.merchantId,
        };
      }

      // 7. Create payment intent
      const paymentIntent = await deps.stripeClient.paymentIntents.create(
        paymentIntentData,
      );
      console.debug(
        'startStripePayment().payment intent created=' + paymentIntent.id,
      );

      // 8. Create Transaction
      const createTransactionVars: CrudApi.CreateTransactionMutationVariables =
        {
          input: {
            userId: userId,
            orderId: orderId,
            currency,
            status: CrudApi.PaymentStatus.waiting_for_payment, // shouldn't we use statusLog instead of the simple actual status ? (Covered by #945)
            externalTransactionId: paymentIntent.id,
            total: amount,
            type: 'stripe',
            paymentMethodId,
          },
        };

      const transaction = await createTransaction(createTransactionVars)(deps);
      console.debug('startStripePayment().transaction.id=' + transaction?.id);

      if (!transaction) {
        throw new Error('Transaction not created');
      }

      await deps.crudSdk
        .UpdateOrder({
          input: {
            id: orderId,
            tip: input.tip,
            tipTransactionStatus: transaction.status,
            tipTransactionId: transaction.id,
          },
        })
        .toPromise();

      console.debug(
        'startStripePayment().updateOrderState.done()=' + order?.id,
      );

      // 11. Return with client secret
      return Promise.resolve({
        clientSecret: paymentIntent.client_secret as string,
        status: paymentIntent.status,
      });
    } else {
      throw new Error('Unsupported tip payment method');
    }
  };
