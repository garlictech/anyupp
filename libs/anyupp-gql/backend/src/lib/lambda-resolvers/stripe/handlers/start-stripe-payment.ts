import * as CrudApi from '@bgap/crud-gql/api';
import Stripe from 'stripe';
import {
  createTransaction,
  loadOrder,
  loadUnit,
  updateOrderState,
} from '../stripe-graphql-crud';
import { StripeResolverDeps } from '../stripe.utils';
import { loadAndConnectUserForStripe } from './common-stripe';
import { toFixed0Number } from '@bgap/shared/utils';
import { createInvoiceAndConnectTransaction } from '../invoice-receipt.utils';

// START PAYMENT INTENTION should use indempotency key https://stripe.com/docs/api/idempotent_requests?lang=node (Covered by #804)
export const startStripePayment =
  (userId: string, input: CrudApi.StartStripePaymentInput) =>
  async (
    deps: StripeResolverDeps,
  ): Promise<CrudApi.StartStripePaymentOutput> => {
    console.debug('startStripePayment().start()');

    // 1. Get parameters, orderId and payment method
    const {
      orderId,
      paymentMethod,
      paymentMethodId,
      savePaymentMethod,
      invoiceAddress,
    } = input;
    console.debug(
      'startStripePayment().orderId=' +
        orderId +
        ', paymentMethod=' +
        paymentMethod +
        ', paymentMethodId=' +
        paymentMethodId +
        ', saveCard=' +
        savePaymentMethod,
    );

    if (paymentMethod == CrudApi.PaymentMethod.inapp && !paymentMethodId) {
      throw Error(
        'Payment method is missing from request when payment mode is INAPP!',
      );
    }
    console.debug('startStripePayment().invoiceAddress=' + invoiceAddress);

    // 2. Load order
    let order = await loadOrder(orderId)(deps);
    console.debug('startStripePayment().order.loaded=' + order?.id);

    if (!order) {
      throw Error('Order not found with id=' + orderId);
    }

    if (userId !== order.userId) {
      throw Error(
        "The Order must belongs to the user, the Order's userId field is not match with the logined User's id! Details: order.userId=" +
          order.userId +
          ' vs userId=' +
          userId,
      );
    }

    const status = order.statusLog[order.statusLog.length - 1];

    if (status.status != CrudApi.OrderStatus.none) {
      throw Error(
        'Order status must be OrderStatus.NONE if you want to pay the order! Current status:' +
          status +
          ', id=' +
          order.id,
      );
    }

    // 3. Load User
    const createStripeCustomer = paymentMethod == CrudApi.PaymentMethod.inapp;
    const user = await loadAndConnectUserForStripe(
      userId,
      createStripeCustomer,
      input.invoiceAddress,
    )(deps);
    console.debug('startStripePayment().user.loaded=' + user?.id);

    if (!user) {
      throw Error(
        'User initialization failed. User not loaded (and maybe not created)!',
      );
    }

    if (
      paymentMethod == CrudApi.PaymentMethod.inapp &&
      !user.stripeCustomerId
    ) {
      throw Error(
        'User initialization failed. User must have a stripeCustomerId property!',
      );
    }

    // 4. Load unit
    console.debug('startStripePayment().loading unit()=' + order.unitId);
    const unit = await loadUnit(order.unitId)(deps);
    console.debug('startStripePayment().unit().loaded=' + unit?.id);
    if (!unit) {
      throw Error('Unit not found with id=' + order.unitId);
    }

    // 5. Handle INAPP payment
    if (paymentMethod == CrudApi.PaymentMethod.inapp) {
      if (!paymentMethodId) {
        throw Error(
          'Payment method is missing from request when payment mode is INAPP!',
        );
      }

      if (!user.stripeCustomerId) {
        throw Error(
          "User don't have a stripeCustomerId property, but it is mandatory for INAPP mayment mode!",
        );
      }

      const stripeAmount: number =
        order.sumPriceShown.currency === 'HUF'
          ? toFixed0Number(order.sumPriceShown.priceSum * 100)
          : order.sumPriceShown.priceSum;

      // 6a. Create payment intent data
      const paymentIntentData: Stripe.PaymentIntentCreateParams = {
        amount: stripeAmount,
        currency: order.sumPriceShown.currency,
        payment_method: paymentMethodId,
        payment_method_types: ['card'],
        customer: user.stripeCustomerId,
      };

      // 6b. Add optional merchantId to the payment
      if (unit?.merchantId) {
        console.debug('startStripePayment().set merchantId=' + unit.merchantId);
        paymentIntentData.application_fee_amount = 0;
        paymentIntentData.transfer_data = {
          destination: unit.merchantId,
        };
      }

      // 6c. Save card for later use
      if (savePaymentMethod === true) {
        await deps.stripeClient.paymentMethods.attach(
          input.paymentMethodId as string,
          {
            customer: user.stripeCustomerId,
          },
        );
      }

      // 7. Create payment intent
      // console.debug('startStripePayment().creating payment intent.');
      const paymentIntent = await deps.stripeClient.paymentIntents.create(
        paymentIntentData,
      );
      console.debug(
        'startStripePayment().payment intent created=' + paymentIntent.id,
      );

      // 8. Create Transaction
      const createTransactionVars: CrudApi.CreateTransactionMutationVariables =
        {
          input: {
            userId: userId,
            orderId: orderId,
            currency: order.sumPriceShown.currency,
            status: CrudApi.PaymentStatus.waiting_for_payment, // shouldn't we use statusLog instead of the simple actual status ? (Covered by #945)
            externalTransactionId: paymentIntent.id,
            total: order.sumPriceShown.priceSum,
            type: 'stripe',
          },
        };
      const transaction = await createTransaction(createTransactionVars)(deps);
      console.debug('startStripePayment().transaction.id=' + transaction?.id);

      if (!transaction) {
        throw new Error('Transaction not created');
      }

      // 9. Create invoice if requested
      if (invoiceAddress) {
        console.debug(
          'startStripePayment().createInvoiceAndConnectTransaction()=' +
            invoiceAddress,
        );
        // Create Invoice
        await createInvoiceAndConnectTransaction(
          order.id,
          order.userId,
          transaction.id,
          invoiceAddress,
          CrudApi.InvoiceStatus.waiting,
        )(deps);
      }

      // 10. Update ORDER STATUS
      order = await updateOrderState(
        order.id,
        userId,
        undefined, // use undefined to prevend graphql api to override this field
        transaction.id,
        // it should be undefined because we don't want to overwrite the field with GraphQL API.
        transaction.status ? transaction.status : undefined,
      )(deps);
      console.debug(
        'startStripePayment().updateOrderState.done()=' + order?.id,
      );

      // 11. Return with client secret
      return Promise.resolve({
        clientSecret: paymentIntent.client_secret as string,
        status: paymentIntent.status,
      });
    } else {
      //
      // Handle CASH and CARD payment
      console.debug('***** startCashPayment()');

      // 6. Create Transaction
      const createTransactionVars: CrudApi.CreateTransactionMutationVariables =
        {
          input: {
            userId: userId,
            orderId: orderId,
            currency: order.sumPriceShown.currency,
            status: CrudApi.PaymentStatus.waiting_for_payment,
            total: order.sumPriceShown.priceSum,
            type: paymentMethod.toString(),
          },
        };

      console.debug('startCashPayment().creating.transaction');
      const transaction = await createTransaction(createTransactionVars)(deps);
      console.debug('startCashPayment().transaction.id=' + transaction?.id);

      if (!transaction) {
        throw new Error('Transaction not created');
      }

      // 7. Create invoice or receipt
      if (invoiceAddress) {
        console.debug(
          'startCashPayment().invoiceAddress=' + input.invoiceAddress,
        );
        await createInvoiceAndConnectTransaction(
          order.id,
          order.userId,
          transaction.id,
          invoiceAddress,
          CrudApi.InvoiceStatus.success,
        )(deps);
      } else {
        // await createReceiptAndConnectTransaction(
        //   order.id,
        //   order.userId,
        //   transaction.id,
        //   user.email,
        //   CrudApi.ReceiptStatus.success,
        // );
      }

      // 8. Update order
      order = await updateOrderState(
        order.id,
        userId,
        undefined, // use undefined to prevend graphql api to override this field
        transaction.id,
        // it should be undefined because we don't want to overwrite the field with GraphQL API.
        transaction.status ? transaction.status : undefined,
      )(deps);
      console.debug('startCashPayment().updateOrderState.done()=' + order?.id);

      // 9. Return with success
      return Promise.resolve({
        clientSecret: '',
        status: CrudApi.PaymentStatus.waiting_for_payment,
      });
    }
  };
