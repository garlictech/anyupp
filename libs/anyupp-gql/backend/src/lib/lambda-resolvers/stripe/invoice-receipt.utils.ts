import * as AnyuppApi from '@bgap/anyupp-gql/api';
import * as CrudApi from '@bgap/crud-gql/api';
import { createInvoice } from './stripe-graphql-crud';
import { StripeResolverDeps } from './stripe.utils';

export const createInvoiceAndConnectTransaction = (
  orderId: string,
  userId: string,
  transactionId: string,
  invoiceAddress: AnyuppApi.UserInvoiceAddress,
  status: CrudApi.InvoiceStatus,
) => async (deps: StripeResolverDeps) => {
  console.debug('createInvoiceAndConnectTransaction().orderId=' + orderId);

  const createInvoiceVars = createInvoiceMutationVariables(
    orderId,
    userId,
    transactionId,
    invoiceAddress,
    status,
  );
  const invoice = await createInvoice(createInvoiceVars)(deps);
  if (!invoice) {
    throw Error('Invoice not created. Created invoice instane is null');
  }
  console.debug('createInvoiceAndConnectTransaction().invoiceId=' + invoice.id);
  await updateTransactionInvoiceId(transactionId, invoice.id)(deps);
};

export const createReceiptAndConnectTransaction = (
  orderId: string,
  userId: string,
  transactionId: string,
  email: string | undefined | null,
  status: CrudApi.ReceiptStatus,
  receiptId: string | undefined,
  recepitPdf: string | undefined,
) => async (deps: StripeResolverDeps) => {
  console.debug('createReceiptAndConnectTransaction().orderId=' + orderId);

  const createReceiptVars = createReceiptMutationVariables(
    orderId,
    userId,
    transactionId,
    email,
    status,
    receiptId,
    recepitPdf,
  );
  const receipt = await deps.crudSdk
    .CreateReceipt(createReceiptVars)
    .toPromise();
  if (!receipt) {
    throw Error('Receipt not created. Created receipt instane is null');
  }
  console.debug('createReceiptAndConnectTransaction().receiptId=' + receipt.id);
  await updateTransactionReceiptId(transactionId, receipt.id)(deps);
};

const createInvoiceMutationVariables = (
  orderId: string,
  userId: string,
  transactionId: string,
  invoiceAddress: AnyuppApi.UserInvoiceAddress,
  status: CrudApi.InvoiceStatus,
): CrudApi.CreateInvoiceMutationVariables => {
  return {
    input: {
      userId,
      orderId,
      transactionId,
      city: invoiceAddress.city,
      country: invoiceAddress.country,
      customerName: invoiceAddress.customerName,
      postalCode: invoiceAddress.postalCode,
      streetAddress: invoiceAddress.streetAddress,
      taxNumber: invoiceAddress.taxNumber,
      email: invoiceAddress.email,
      status,
    },
  };
};

const createReceiptMutationVariables = (
  orderId: string,
  userId: string,
  transactionId: string,
  email: string | undefined | null,
  status: CrudApi.ReceiptStatus,
  receiptId: string | undefined,
  pdfData: string | undefined,
): CrudApi.CreateReceiptMutationVariables => {
  return {
    input: {
      orderId,
      userId,
      transactionId,
      email,
      status,
      pdfData,
      externalReceiptId: receiptId,
    },
  };
};

const updateTransactionInvoiceId = (id: string, invoiceId: string) => (
  deps: StripeResolverDeps,
) => {
  const updateTransactionVars: CrudApi.UpdateTransactionMutationVariables = {
    input: {
      id,
      invoiceId,
    },
  };

  return deps.crudSdk.UpdateTransaction(updateTransactionVars).toPromise();
};

const updateTransactionReceiptId = (id: string, receiptId: string) => (
  deps: StripeResolverDeps,
) => {
  const updateTransactionVars: CrudApi.UpdateTransactionMutationVariables = {
    input: {
      id,
      receiptId,
    },
  };

  return deps.crudSdk.UpdateTransaction(updateTransactionVars).toPromise();
};
