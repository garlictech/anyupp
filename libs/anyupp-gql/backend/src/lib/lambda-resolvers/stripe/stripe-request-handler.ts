import { AnyuppApi } from '@bgap/anyupp-gql/api';
import { missingParametersCheck } from '@bgap/shared/utils';
import * as stripeService from './stripe.service';
import { GraphqlApiClient } from '@bgap/shared/graphql/api-client';

interface WithStripeCustomer {
  stripeCustomerId: string;
}
interface WithCognitoUser {
  userId: string;
}

type UpdateStripeCardRequest = WithStripeCustomer &
  AnyuppApi.MutationUpdateMyStripeCardArgs;
type DeleteStripeCardRequest = WithStripeCustomer &
  AnyuppApi.MutationDeleteMyStripeCardArgs;
type StartStripePaymentRequest = WithCognitoUser &
  AnyuppApi.MutationStartStripePaymentArgs;

export type ListStripeCardsRequest = WithCognitoUser;

export const stripeRequestHandler = {
  // LIST STRIPE CARDS FOR THE LOGGED IN CUSTOMER
  listStripeCards: (crudGraphqlClient: GraphqlApiClient) => (requestPayload: ListStripeCardsRequest,
  ): Promise<AnyuppApi.StripeCard[]> => {

    return stripeService.listStripeCards(crudGraphqlClient, requestPayload.userId);
  },

  startStripePayment: (crudGraphqlClient: GraphqlApiClient) => (requestPayload: StartStripePaymentRequest,
    ): Promise<AnyuppApi.StartStripePaymentOutput> => { // TODO
    const {
      userId,
      input,
    } = requestPayload as StartStripePaymentRequest;

    // missingParametersCheck(requestPayload, ['stripeCustomerId']);
    // missingParametersCheck(input, [
    //   'chainId',
    //   'unitId',
    //   'paymentMethod',
    //   'userLocation',
    // ]);

    return stripeService.startStripePayment(crudGraphqlClient, userId, input);
  },

  getStripeCardsForCustomer(
    requestPayload: unknown,
  ): Promise<AnyuppApi.StripeCard[]> {
    const { stripeCustomerId } = requestPayload as WithStripeCustomer;

    if (!stripeCustomerId) {
      return Promise.resolve([]);
    }

    return stripeService.getStripeCardsForCustomer(stripeCustomerId);
  },
  updateStripeCard(requestPayload: unknown): Promise<AnyuppApi.StripeCard> {
    const {
      stripeCustomerId,
      input,
    } = requestPayload as UpdateStripeCardRequest;

    missingParametersCheck(requestPayload, ['stripeCustomerId']);
    missingParametersCheck(input, ['id']);

    return stripeService.updateStripeCard(stripeCustomerId, input);
  },
  deleteStripeCard(requestPayload: unknown) {
    const {
      stripeCustomerId,
      input,
    } = requestPayload as DeleteStripeCardRequest;

    missingParametersCheck(requestPayload, ['stripeCustomerId']);
    missingParametersCheck(input, ['id']);

    return stripeService.deleteStripeCard(stripeCustomerId, input);
  },
  startStripePaymentOrig(requestPayload: unknown) {
    const {
      stripeCustomerId,
      input,
    } = requestPayload as StartStripePaymentRequest;

    missingParametersCheck(requestPayload, ['stripeCustomerId']);
    missingParametersCheck(input, [
      'chainId',
      'unitId',
      'paymentMethod',
      'userLocation',
    ]);

    return stripeService.startStripePayment(stripeCustomerId, input);
  },
};
