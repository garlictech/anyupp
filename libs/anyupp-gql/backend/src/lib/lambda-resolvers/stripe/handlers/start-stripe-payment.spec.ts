/* eslint-disable @typescript-eslint/no-explicit-any */
import { startStripePayment } from './start-stripe-payment';
import { payTipWithStripe } from './pay-tip-with-stripe';
import * as CrudApi from '@bgap/crud-gql/api';
import { StripeResolverDeps } from '../stripe.utils';
import { of } from 'rxjs';

const getDepsMock = (): StripeResolverDeps => ({
  userId: 'USER_ID',
  crudSdk: {
    GetOrder: jest.fn().mockReturnValue(
      of({
        id: 'ORDER_ID',
        items: [],
        unitId: 'UNIT_ID',
        userId: 'USER_ID',
        statusLog: [
          {
            userId: 'USER_ID',
            status: CrudApi.OrderStatus.none,
            ts: 123456,
          },
        ],
        paymentMode: {
          method: CrudApi.PaymentMethod.inapp,
          type: CrudApi.PaymentType.stripe,
        },
        sumPriceShown: {
          currency: 'BATKA',
          priceSum: 100,
        },
        transaction: {
          paymentMethodId: 'TRANSACTION PAYMENT METHOD ID',
        },
        serviceFee: {
          currency: 'BATKA',
          grossPrice: 100,
          taxContent: 10,
        },
        packagingSum: {
          currency: 'BATKA',
          netPrice: 10,
          taxPercentage: 20,
        },
      }),
    ),
    GetUser: jest.fn().mockReturnValue(
      of<CrudApi.User>({
        id: 'USER_ID',
        createdAt: 'CREATEDAT',
        updatedAt: 'UPDATEDAT',
      }),
    ),
    UpdateUser: jest
      .fn()
      .mockReturnValue(of({ stripeCustomerId: 'STRIPE_CUSTOMER_ID' })),
    GetUnit: jest.fn().mockReturnValue(of({ id: 'UNIT_ID' })),
    CreateTransaction: jest.fn().mockReturnValue(
      of({
        id: 'TRANSACTION_ID',
        status: CrudApi.PaymentStatus.waiting_for_payment,
      }),
    ),
    UpdateTransaction: jest.fn().mockReturnValue(
      of({
        id: 'TRANSACTION_ID',
      }),
    ),
    UpdateOrder: jest.fn().mockReturnValue(
      of({
        id: 'ORDER_ID',
      }),
    ),
    CreateInvoice: jest.fn().mockReturnValue(
      of({
        id: 'INVOICE_ID',
      }),
    ),
  } as any,
  szamlazzClient: {} as any,
  stripeClient: {
    customers: {
      create: jest.fn().mockReturnValue(
        Promise.resolve({
          id: 'STRIPE_CUSTOMER_ID',
        }),
      ),
    },
    paymentIntents: {
      create: jest.fn().mockReturnValue(
        Promise.resolve({
          id: 'STRIPE PAYMENT INTENT ID',
          client_secret: 'STRIPE CLIENT SECRET',
          status: 'processing',
        }),
      ),
    },
  } as any,
});

const checkMocks = (deps: StripeResolverDeps) => {
  Object.keys(deps.crudSdk).forEach(key =>
    expect((deps.crudSdk as any)[key].mock.calls).toMatchSnapshot(
      `${key} calls`,
    ),
  );

  expect(
    (deps.stripeClient.customers.create as any).mock.calls,
  ).toMatchSnapshot('stripeClient.customers.create.mock');

  expect(
    (deps.stripeClient.paymentIntents.create as any).mock.calls,
  ).toMatchSnapshot('stripeClient.paymentIntents.create.mock');
};

test('startStripePayment test', async () => {
  const input: CrudApi.StartStripePaymentInput = {
    invoiceAddress: {
      city: 'Pornóapáti',
      country: 'Neverland',
      customerName: 'Test Elek',
      postalCode: '1234',
      streetAddress: 'Netuddhol',
      taxNumber: '123456',
    },
    orderId: 'ORDER_ID',
    paymentMethod: CrudApi.PaymentMethod.inapp,
    savePaymentMethod: false,
    paymentMethodId: 'PAYMENT_METHOD_ID',
  };

  const deps = getDepsMock();
  const res = await startStripePayment(input)(deps);

  expect(res).toMatchSnapshot('Stripe payment result');
  checkMocks(deps);
});

test('payTipWithStripe test - fixed amount', async () => {
  const input: CrudApi.PayTipWithStripeInput = {
    orderId: 'ORDER_ID',
    tip: {
      value: 100,
      type: CrudApi.TipType.amount,
    },
  };

  const deps = getDepsMock();
  const res = await payTipWithStripe(input)(deps);

  expect(res).toMatchSnapshot('Stripe tip payment result');
  checkMocks(deps);
});

test('payTipWithStripe test - percentage', async () => {
  const input: CrudApi.PayTipWithStripeInput = {
    orderId: 'ORDER_ID',
    tip: {
      value: 10,
      type: CrudApi.TipType.percent,
    },
  };

  const deps = getDepsMock();
  const res = await payTipWithStripe(input)(deps);

  expect(res).toMatchSnapshot('Stripe tip payment result');
  checkMocks(deps);
});
