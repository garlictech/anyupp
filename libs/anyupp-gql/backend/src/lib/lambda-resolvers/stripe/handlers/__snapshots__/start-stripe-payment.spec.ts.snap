// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payTipWithStripe test: CreateInvoice calls 1`] = `Array []`;

exports[`payTipWithStripe test: CreateTransaction calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "currency": "BATKA",
        "externalTransactionId": "STRIPE PAYMENT INTENT ID",
        "orderId": "ORDER_ID",
        "paymentMethodId": "TRANSACTION PAYMENT METHOD ID",
        "status": "waiting_for_payment",
        "total": 10,
        "type": "stripe",
        "userId": "USER_ID",
      },
    },
  ],
]
`;

exports[`payTipWithStripe test: GetOrder calls 1`] = `
Array [
  Array [
    Object {
      "id": "ORDER_ID",
    },
    Object {
      "fetchPolicy": "network-only",
    },
  ],
]
`;

exports[`payTipWithStripe test: GetUnit calls 1`] = `
Array [
  Array [
    Object {
      "id": "UNIT_ID",
    },
    Object {
      "fetchPolicy": "network-only",
    },
  ],
]
`;

exports[`payTipWithStripe test: GetUser calls 1`] = `
Array [
  Array [
    Object {
      "id": "USER_ID",
    },
    Object {
      "fetchPolicy": "network-only",
    },
  ],
]
`;

exports[`payTipWithStripe test: Stripe tip payment result 1`] = `
Object {
  "clientSecret": "STRIPE CLIENT SECRET",
  "status": "processing",
}
`;

exports[`payTipWithStripe test: UpdateOrder calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "id": "ORDER_ID",
        "tip": Object {
          "type": "amount",
          "value": 10,
        },
        "tipTransactionId": "TRANSACTION_ID",
        "tipTransactionStatus": "waiting_for_payment",
      },
    },
  ],
]
`;

exports[`payTipWithStripe test: UpdateTransaction calls 1`] = `Array []`;

exports[`payTipWithStripe test: UpdateUser calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "id": "USER_ID",
        "invoiceAddress": undefined,
        "stripeCustomerId": "STRIPE_CUSTOMER_ID",
      },
    },
  ],
]
`;

exports[`payTipWithStripe test: stripeClient.customers.create.mock 1`] = `
Array [
  Array [],
]
`;

exports[`payTipWithStripe test: stripeClient.paymentIntents.create.mock 1`] = `
Array [
  Array [
    Object {
      "amount": 10,
      "currency": "BATKA",
      "customer": "STRIPE_CUSTOMER_ID",
      "payment_method": "TRANSACTION PAYMENT METHOD ID",
      "payment_method_types": Array [
        "card",
      ],
      "statement_descriptor_suffix": "tip",
    },
  ],
]
`;

exports[`startStripePayment test: CreateInvoice calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "city": "Porn칩ap치ti",
        "country": "Neverland",
        "customerName": "Test Elek",
        "email": undefined,
        "orderId": "ORDER_ID",
        "postalCode": "1234",
        "status": "waiting",
        "streetAddress": "Netuddhol",
        "taxNumber": "123456",
        "transactionId": "TRANSACTION_ID",
        "userId": "USER_ID",
      },
    },
  ],
]
`;

exports[`startStripePayment test: CreateTransaction calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "currency": "BATKA",
        "externalTransactionId": "STRIPE PAYMENT INTENT ID",
        "orderId": "ORDER_ID",
        "paymentMethodId": "PAYMENT_METHOD_ID",
        "status": "waiting_for_payment",
        "total": 100,
        "type": "stripe",
        "userId": "USER_ID",
      },
    },
  ],
]
`;

exports[`startStripePayment test: GetOrder calls 1`] = `
Array [
  Array [
    Object {
      "id": "ORDER_ID",
    },
    Object {
      "fetchPolicy": "network-only",
    },
  ],
  Array [
    Object {
      "id": "ORDER_ID",
    },
    Object {
      "fetchPolicy": "network-only",
    },
  ],
]
`;

exports[`startStripePayment test: GetUnit calls 1`] = `
Array [
  Array [
    Object {
      "id": "UNIT_ID",
    },
    Object {
      "fetchPolicy": "network-only",
    },
  ],
]
`;

exports[`startStripePayment test: GetUser calls 1`] = `
Array [
  Array [
    Object {
      "id": "USER_ID",
    },
    Object {
      "fetchPolicy": "network-only",
    },
  ],
]
`;

exports[`startStripePayment test: Stripe payment result 1`] = `
Object {
  "clientSecret": "STRIPE CLIENT SECRET",
  "status": "processing",
}
`;

exports[`startStripePayment test: UpdateOrder calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "id": "ORDER_ID",
        "items": Array [],
        "statusLog": undefined,
        "transactionId": "TRANSACTION_ID",
        "transactionStatus": "waiting_for_payment",
      },
    },
  ],
]
`;

exports[`startStripePayment test: UpdateTransaction calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "id": "TRANSACTION_ID",
        "invoiceId": "INVOICE_ID",
      },
    },
  ],
]
`;

exports[`startStripePayment test: UpdateUser calls 1`] = `
Array [
  Array [
    Object {
      "input": Object {
        "id": "USER_ID",
        "invoiceAddress": Object {
          "city": "Porn칩ap치ti",
          "country": "Neverland",
          "customerName": "Test Elek",
          "email": undefined,
          "postalCode": "1234",
          "streetAddress": "Netuddhol",
          "taxNumber": "123456",
        },
        "stripeCustomerId": "STRIPE_CUSTOMER_ID",
      },
    },
  ],
]
`;

exports[`startStripePayment test: stripeClient.customers.create.mock 1`] = `
Array [
  Array [],
]
`;

exports[`startStripePayment test: stripeClient.paymentIntents.create.mock 1`] = `
Array [
  Array [
    Object {
      "amount": 100,
      "currency": "BATKA",
      "customer": "STRIPE_CUSTOMER_ID",
      "payment_method": "PAYMENT_METHOD_ID",
      "payment_method_types": Array [
        "card",
      ],
    },
  ],
]
`;
